substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/jdk11/ClassLoaderSupportImplJDK11OrLater.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2021, 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.hosted.jdk11;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.MissingResourceException;
+import java.util.ResourceBundle;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.graalvm.compiler.serviceprovider.JavaVersionUtil;
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.hosted.Feature;
+
+import com.oracle.svm.core.ClassLoaderSupport;
+import com.oracle.svm.core.annotate.AutomaticFeature;
+import com.oracle.svm.hosted.FeatureImpl;
+import com.oracle.svm.hosted.NativeImageClassLoaderSupport;
+import com.oracle.svm.hosted.NativeImageSystemClassLoader;
+
+import jdk.internal.module.Modules;
+
+public final class ClassLoaderSupportImplJDK11OrLater extends ClassLoaderSupport {
+
+    private final NativeImageClassLoaderSupport classLoaderSupport;
+    private final Map<String, Set<Module>> packageToModules;
+
+    ClassLoaderSupportImplJDK11OrLater(NativeImageClassLoaderSupport classLoaderSupport) {
+        this.classLoaderSupport = classLoaderSupport;
+        packageToModules = new HashMap<>();
+        buildPackageToModulesMap(classLoaderSupport);
+    }
+
+    @Override
+    protected boolean isNativeImageClassLoaderImpl(ClassLoader loader) {
+        return loader == classLoaderSupport.getClassLoader() || loader instanceof NativeImageSystemClassLoader;
+    }
+
+    @Override
+    public List<ResourceBundle> getResourceBundle(String bundleSpec, Locale locale) {
+        String[] specParts = bundleSpec.split(":", 2);
```

Comment:
Clever idea to allow the user to specify the module from which the bundle should be loaded. However, I would put the parsing into the caller and have the module as an optional argument. For the command-line options, it makes sense to have this parsing, but for the config file, a new key-value pair separate from the bundle name might be more expressive. And the classloader support shouldn't be aware of the nuances of locale configuration. However, this is more of a `todo` for me to implement it once this gets merged.

Reply:
> Clever idea to allow the user to specify the module from which the bundle should be loaded.

Thanks. The idea is actually from @ilopmar https://github.com/oracle/graal/issues/3404#issuecomment-839888673
This is what he naturally tried to use (expecting it to work). I just identified it as being worth to be implemented.